### PR TITLE
Update the design of the 'Advanced' button

### PR DESCRIPTION
**Pull request by janfaracik**
This PR updates the design of the 'Advanced' button. It now uses the new button classes, has a new icon and is now toggle able (rather than disappearing).

**Before**
image

**After**
image

image

### Proposed changelog entries

- Update the design of the 'Advanced' button

### Proposed upgrade guidelines

N/A

### Developer upgrade guidelines

- `align` and `style` attributes have been removed from the `<f:advanced />` component

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7173"></a>

### CODE REVIEW

**Review by timja — COMMENTED** [review]
I like it, needs ATH testing though

**Comment by daniel-beck** [comment]
This doesn't work in conjunction with the marker for settings that have been edited. Since the marker and button remain after being expanded, why wouldn't the marker update with the current state of all subordinate options? Especially after closing the section again, the presence/absence of the marker is actively misleading. Previously, the marker and button disappear after expanding, so this wasn't even a problem.

Also, design wise, there is no distance between the button and the marker anymore, which simply looks bad.
Screenshot before
Screenshot after
Screenshot after with focus

**Comment by NotMyFault** [comment]
ATH fails: https://github.com/jenkinsci/acceptance-test-harness/pull/955

**Comment by janfaracik** [comment]
> This PR also currently breaks whatever recognizes that an "Advanced…" button click does not constistute a change that might need saving, so navigating away or reloading the page would not ask to save unchanged changes.
> 
> Also, this removes multiple options from a tag without documenting their removal.
> 
> Also, design wise, there is no distance between the button and the marker anymore, which simply looks bad.

Thanks for raising - all good spots. I've addressed these in the last couple of commits. 

I'll take a look at `> This doesn't work in conjunction with the marker for settings that have been edited.`.

**Comment by janfaracik** [comment]
The 'Edited' marker will now automatically change visibility when a field is edited. 

It doesn't change visibility however if you load a form that's already been changed.

https://user-images.githubusercontent.com/43062514/193408491-0387ce51-7bba-4a33-a995-6c5b1e9327b3.mov

Hope that's alright! Let me know.

**Comment by daniel-beck** [comment]
@janfaracik Thanks, the demo looks awesome! The indicator is a lot clearer. I'm looking forward to testing this.

**Review by yaroslavafenkin — COMMENTED** [review]
Tested a little, generally looks good, but I found some bugs.

- Weird content shift on freestyle job configuration page. Here's a recording

https://user-images.githubusercontent.com/91559310/194866455-a87285ed-1b73-41a3-9edd-8659aa65dcc6.mov

It shifts on its own, it's not me manually scrolling upon hitting the checkbox :p

- On clean instance after installing `mailer-plugin` and going to Global Configuration page, scrolling down to E-mail Notification section, there will be _Edited_ label next to advanced button while there are no changes.

**Comment by janfaracik** [comment]
> It shifts on its own, it's not me manually scrolling upon hitting the checkbox

Ah that'll be preexisting and somethings that I've found annoying lots of times. Opened a PR to remove that.

> * On clean instance after installing `mailer-plugin` and going to Global Configuration page, scrolling down to E-mail Notification section, there will be _Edited_ label next to advanced button while there are no changes.

I'll investigate - thanks!

**Comment by daniel-beck** [comment]
>     * On clean instance after installing `mailer-plugin` and going to Global Configuration page, scrolling down to E-mail Notification section, there will be _Edited_ label next to advanced button while there are no changes.

Make sure this is new in this PR, it might not be. Some plugins have weird config forms (and we should consider it a bug in the plugin).

**Comment by timja** [comment]
> Make sure this is new in this PR, it might not be. Some plugins have weird config forms (and we should consider it a bug in the plugin).

This is on the master branch:
image

Not introduced as part of this pull request

**Comment by timja** [comment]
Tests are failing FYI @janfaracik

**Comment by timja** [comment]
still some failing

**Comment by daniel-beck** [comment]
https://github.com/jenkinsci/jenkins/blob/03c380f6c451a1a05e3b0a8471f3aa06a24c3baa/test/src/test/java/hudson/model/ProjectTest.java#L266 fails

**Review by daniel-beck — DISMISSED** [review]
This does not interact properly with the "X Installations…" button on Global Tool Configuration when you have at least 1 installation for a given tool defined. Previously, the button disappears when clicked, but in this PR it does not. Clicking it again does not collapse the list of tool installations either, and there's no collapsed/expanded indicator.

**Comment by janfaracik** [comment]
> This does not interact properly with the "X Installations…" button on Global Tool Configuration when you have at least 1 installation for a given tool defined. Previously, the button disappears when clicked, but in this PR it does not. Clicking it again does not collapse the list of tool installations either, and there's no collapsed/expanded indicator.

Great spot! That was a custom implementation of the button, I've replaced it with the actual component now - from my understanding it looks like it was a custom implementation due to layout implications, which aren't an issue anymore.

image

Would be grand if you could check it out when you're free, thanks.

**Comment by daniel-beck** [comment]
> That was a custom implementation of the button, I've replaced it with the actual component now - from my understanding it looks like it was a custom implementation due to layout implications, which aren't an issue anymore.

Did you check plugins for other occurrences of this pattern? If not, could you do that?

**Comment by scherler** [comment]
I wrote my observation in detail in the ux-sig channel, but bottom line I would love to see that we retire the hudson-behavior function and instead use an event driven approach to see the extra information. I strongly belive the hudson-behavior must die.

**Comment by scherler** [comment]
I created an example https://github.com/jenkinsci/jenkins/compare/master...scherler:web_component_checkbox?expand=1#diff-a876c43d1646d94029c38d2398130e24a70754fe924e7ea735c94f2e896144c8R83

**Comment by janfaracik** [comment]
> > That was a custom implementation of the button, I've replaced it with the actual component now - from my understanding it looks like it was a custom implementation due to layout implications, which aren't an issue anymore.
> 
> Did you check plugins for other occurrences of this pattern? If not, could you do that?

Did a search for `class="advancedLink"` and found:

https://github.com/jenkinsci/vstart-plugin (7 years since the last commit)

And also a reference to 'advancedLink' in `expandButton.jelly` in core Jenkins which from what I've seen is only used in https://github.com/jenkinsci/javanet-plugin (11 years since the last commit).

**Comment by timja** [comment]
Also this: https://github.com/jenkinsci/environment-dashboard-plugin/blob/389a87fafac0eb662a462dfd9d9e37dcc6a58a0b/src/main/resources/org/jenkinsci/plugins/environmentdashboard/EnvDashboardView/index.jelly#L12

might be worth looking at

**Comment by janfaracik** [comment]
> Also this: https://github.com/jenkinsci/environment-dashboard-plugin/blob/389a87fafac0eb662a462dfd9d9e37dcc6a58a0b/src/main/resources/org/jenkinsci/plugins/environmentdashboard/EnvDashboardView/index.jelly#L12
> 
> might be worth looking at

I've just tried it and it seems to be working fine 👍 (although very much right aligned)

image

**Comment by daniel-beck** [comment]
> https://github.com/jenkinsci/vstart-plugin (7 years since the last commit)

Doesn't look like that has ever been properly released. I archived it.

> in https://github.com/jenkinsci/javanet-plugin (11 years since the last commit).

Unsure whether that is https://plugins.jenkins.io/javanet/ or https://plugins.jenkins.io/javanet-uploader/ but both are suspended and have been for a long time. I also archived this repo.

> And also a reference to 'advancedLink' in `expandButton.jelly` in core Jenkins

Does https://github.com/jenkinsci/jenkins/blob/d3d1c88d299735186209ad5b2f7a41e05486105f/core/src/main/resources/lib/layout/expandButton.jelly#L35-L37 need to be adapted?

File: test/src/test/java/lib/form/HeteroListTest.java
Comment (daniel-beck) [inline test/src/test/java/lib/form/HeteroListTest.java:165]: Why is this change needed?

File: test/src/test/java/lib/form/AdvancedButtonTest.java
Comment (daniel-beck) [inline test/src/test/java/lib/form/AdvancedButtonTest.java:34]: Is it not useful to use `…` as indicator that clicking does not result in an immediate change?

File: test/src/test/java/lib/form/HeteroListTest.java
Comment (timja) [reply test/src/test/java/lib/form/HeteroListTest.java:165]: There's an svg in there now as well

File: test/src/test/java/lib/form/AdvancedButtonTest.java
Comment (janfaracik) [reply test/src/test/java/lib/form/AdvancedButtonTest.java:34]: There's a small icon in the button now rather than the three dots to let the user know that content will appear below.

**Comment by janfaracik** [comment]
> need to be adapted, or does it work as it needs to?

Yep still works 👍

**Comment by timja** [comment]
ATH still needs to go green

**Comment by janfaracik** [comment]
> ATH still needs to go green

On it

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Nice work done

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
The integration of this PR has caused two new test failures in `hudson.security.LDAPSecurityRealmTest#groupMembershipAttribute` and `hudson.plugins.emailext.ExtendedEmailPublisherDescriptorTest#defaultTriggers`.

**Comment by timja** [comment]
> The integration of this PR has caused two new test failures in `hudson.security.LDAPSecurityRealmTest#groupMembershipAttribute` and `hudson.plugins.emailext.ExtendedEmailPublisherDescriptorTest#defaultTriggers`.

Thanks, fixed in:

- https://github.com/jenkinsci/email-ext-plugin/pull/414
- https://github.com/jenkinsci/ldap-plugin/pull/211

**Pull request merged**
